File: getting-started-mocks/09-when-handwritten-mocks-stop-scaling/node/notifier/recorder.js
// A fake: a working, stateful implementation of the recorder seam.
export class MemoryRecorder {
  #messages = new Map()

  async record(trackingNumber, message) {
    if (!this.#messages.has(trackingNumber)) {
      this.#messages.set(trackingNumber, message)
    }
  }

  async notified(trackingNumber) {
    return this.#messages.has(trackingNumber)
  }

  message(trackingNumber) {
    return this.#messages.get(trackingNumber) ?? null
  }

  get count() {
    return this.#messages.size
  }
}

File: getting-started-mocks/09-when-handwritten-mocks-stop-scaling/python/notifier/recorder.py
class MemoryRecorder:
    """A fake: a working, stateful implementation of the recorder seam."""

    def __init__(self) -> None:
        self._messages: dict[str, str] = {}

    def record(self, tracking_number: str, message: str) -> None:
        self._messages.setdefault(tracking_number, message)

    def notified(self, tracking_number: str) -> bool:
        return tracking_number in self._messages

    def message(self, tracking_number: str) -> str | None:
        return self._messages.get(tracking_number)

    @property
    def count(self) -> int:
        return len(self._messages)
